### PR TITLE
Pykello/slice

### PR DESCRIPTION
**Pull request by macieksarnowicz**
Adding ResourceGroup object to track systemctl slice and control resources on the host

Adding a new model + prog for ResourceGroup. A ResourceGroup controls the allocation of CPUs for now, and it can be expanded in the future to control other resources. When created in Clover and associated with a VmHost it creates a SystemCtl slice object (and underlying cgroup) on the host. It takes as an argument a set of CPUs to be affinitized to the group. The format we use is the same as used by the OS, with a ranges of CPUs separated by commas, for example: "2-3,6-7" or "4,6-7", etc. The Prog:Vm:ResourceGroup controls the lifecycle of a host-side slice, with prep, wait, and destroy states. Upon host restart, we perform necessary steps to also restart the slice.
To create a ResourceGroup:
Prog::Vm::ResourceGroup.assemble_with_host("standard123", VmHost.first, allowed_cpus: "2-3", memory_1g: 8)

VmHost can list all available resource groups.

Added tests with 100% coverage.

### CODE REVIEW

File: demo/cloudify_server [hunk 1]
```diff
@@ -74,6 +80,13 @@ begin
       end
       sleep 2
     end
+
+    puts "Downloading the default boot image - Ubuntu Jammy 22.04"
+    download = strand.subject.download_boot_image("ubuntu-jammy", version: "20240319")
+    while download.exists?
+      sleep 2
+      print "."
+    end
```
Comment (pykello) [inline demo/cloudify_server]: Downloading a boot image when preparing a server can also be done by adding `default_boot_images: ["ubuntu-jammy"]` to the parameters of `Prog::Vm::HostNexus.assemble`. It'll automatically use the image version from `Config.ubuntu_jammy_version` which is set in `config.rb`.

File: migrate/20241108_resource_group.rb [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+      column :id, :uuid, primary_key: true
+      column :name, :text, null: false
+      column :allocation_state, :allocation_state, default: "unprepared", null: false
+      column :type, :resource_group_type, default: "dedicated", null: false
+      column :allowed_cpus, :text, null: false
+      column :cores, Integer, null: false
+      column :total_cpu_percent, Integer, null: false
+      column :used_cpu_percent, Integer, null: false
+      column :total_memory_1g, Integer, null: false
+      column :used_memory_1g, Integer, null: false
```
Comment (pykello) [inline migrate/20241108_resource_group.rb]: If we aren't doing memory allocation using slices for now, I suggest we remove these. These can be added later if needed.

File: migrate/20241108_resource_group.rb [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+    create_table(:resource_group) do
+      column :id, :uuid, primary_key: true
+      column :name, :text, null: false
+      column :allocation_state, :allocation_state, default: "unprepared", null: false
+      column :type, :resource_group_type, default: "dedicated", null: false
+      column :allowed_cpus, :text, null: false
+      column :cores, Integer, null: false
+      column :total_cpu_percent, Integer, null: false
+      column :used_cpu_percent, Integer, null: false
```
Comment (pykello) [inline migrate/20241108_resource_group.rb]: How are these fields (cores, total_cpu_percent, used_cpu_percent) will be used in future commits?

File: prog/vm/resource_group.rb [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        cores: 0,
+        total_cpu_percent: 0,
+        used_cpu_percent: 0,
+        total_memory_1g: memory_1g,
+        used_memory_1g: 0,
+        vm_host_id: vm_host.id
+        ) { _1.id = ubid.to_uuid }
+
+      # This validates the cpuset and updates or related values
+      rg.from_cpu_bitmask(ResourceGroup.cpuset_to_bitmask(allowed_cpus))
```
Comment (pykello) [inline prog/vm/resource_group.rb]: If `from_cpu_bitmask` and `cpuset_to_bitmask` are only there to enable validation, what do you think about writing a single simpler `ResourceGroup.validate` function instead of multiple functions? If we do that, I think we can implement something simpler without using the `bitarray` gem.

Then it can be called before the `DB.transaction do` line.

File: prog/vm/host_nexus.rb [hunk 5]
```diff
@@ -183,6 +183,14 @@ def before_run
...
 
+    hop_start_resource_groups
+  end
+
+  label def start_resource_groups
+    vm_host.resource_groups.each { |rg|
+      rg.incr_start_after_host_reboot
+    }
+
     hop_start_vms
```
Comment (pykello) [inline prog/vm/host_nexus.rb:194]: Will starting VMs require that slices have been properly setup as a prerequisite? Or can they be done concurrently?

If the ordering doesn't matter, then the current code is fine.

Otherwise, we need to wait until the slice setup has finished before jumping to the next step (or do a synchronous call from here to a new method in ResourceGroup which calls the shell command to setup the slice).

File: migrate/20241108_resource_group.rb [hunk 6]
```diff
@@ -0,0 +1,29 @@
+# frozen_string_literal: true
```
Comment (pykello) [inline migrate/20241108_resource_group.rb]: The migration files (+ changes to `*.cache`)  are usually done in a separate commit than the changes to the files in the `model` directory.

@fdr can you comment on what the reason was?

File: migrate/20241108_resource_group.rb [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+
+Sequel.migration do
+  change do
+    create_enum(:resource_group_type, %w[dedicated shared])
+
+    create_table(:resource_group) do
+      column :id, :uuid, primary_key: true
+      column :name, :text, null: false
+      column :allocation_state, :allocation_state, default: "unprepared", null: false
```
Comment (pykello) [inline migrate/20241108_resource_group.rb]: I'm a bit unsure about if we should use the `allocation_state` enum here, as it's tied a bit to tracking state of a VmHost. Although right now this seems fine, but using it for a 2nd purpose might make changing it more difficult in the future.

Can we use an `enabled` boolean instead?

**Review by pykello — COMMENTED** [review]
Thanks for the PR!

I think the `demo/cloudify_server` changes should be in a separate PR against the main ubicloud repository.

File: demo/cloudify_server (hunk 1 above)
Comment (macieksarnowicz) [reply demo/cloudify_server]: Fair. I can remove it from this PR and send a separate one, and also consolidate the image download into the main call.

File: migrate/20241108_resource_group.rb (hunk 2 above)
Comment (macieksarnowicz) [reply migrate/20241108_resource_group.rb]: This will be needed for the allocator. The VmHost will "give" memory to the slice and then subtract from it for each VM allocated in the slice. I will need this in subsequent PR and I just wanted to minimize the number of migrations.

File: migrate/20241108_resource_group.rb (hunk 3 above)
Comment (macieksarnowicz) [reply migrate/20241108_resource_group.rb]: Similar to memory. Allocator uses Cores, but then VMs will use CPU percentage. Those values are calculated from the cpuset mask that is passed to the ResourceGroup.

File: prog/vm/resource_group.rb (hunk 4 above)
Comment (macieksarnowicz) [reply prog/vm/resource_group.rb]: They are not just for validation. We use AllowedCPU string format as storage and to pass to systemd calls, as that's what it uses. It is easier to store in that format. At the same time, that format is hard to compare with other cpusets and do any calculations on, so I want to use the BitArray for that. This will let Allocator and other logic to merge BitArrays and look for not used CPU ranges.

File: prog/vm/host_nexus.rb (hunk 5 above)
Comment (macieksarnowicz) [reply prog/vm/host_nexus.rb:194]: In here, the slice is already set up, i.e. it is present in the system. The RG.incr_start_after_host_reboot does two things: 
- starts the slice
- re-sets the "root" property
Starting should be done before the VM starts, but VM has a dependency on the slice, so the system will ensure the right order. The "root" property can be set independently.

File: migrate/20241108_resource_group.rb (hunk 6 above)
Comment (macieksarnowicz) [reply migrate/20241108_resource_group.rb]: We are discussing this in slack. I completely agree that we need to have migrations deployed separately, but I think we will finish this work first in this repo and then move it over to ubicloud/main and at that time we will do that separation. We will know the full extent of migration changes then.

File: migrate/20241108_resource_group.rb (hunk 7 above)
Comment (macieksarnowicz) [reply migrate/20241108_resource_group.rb]: OK, fair. I just wanted to not invent a new enum. I can change it to boolean and we will see if we need more than 2 states and come back to it, if needed.
